front/pages/api/w/[wId]/workspace-analytics.ts

```diff
@@ -0,0 +1,237 @@
+import type { NextApiRequest, NextApiResponse } from "next";
+import { QueryTypes } from "sequelize";
+
+import { Authenticator, getSession } from "@app/lib/auth";
+import { front_sequelize } from "@app/lib/databases";
+import { apiError, withLogging } from "@app/logger/withlogging";
+
+export type GetWorkspaceAnalyticsResponse = {
+  memberCount: number;
+  monthlyActiveUsers: {
+    count: number;
+    growth: number;
+  };
+  weeklyActiveUsers: {
+    count: number;
+    growth: number;
+  };
+  averageWeeklyDailyActiveUsers: {
+    count: number;
+    growth: number;
+  };
+};
+
+async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse<GetWorkspaceAnalyticsResponse | APIErrorResponse>
+): Promise<void> {
+  const session = await getSession(req, res);
+  const auth = await Authenticator.fromSession(
+    session,
+    req.query.wId as string
+  );
+
+  const owner = auth.workspace();
+  if (!owner) {
+    return apiError(req, res, {
+      status_code: 404,
+      api_error: {
+        type: "workspace_not_found",
+        message: "The workspace was not found.",
+      },
+    });
+  }
+
+  if (!auth.isAdmin()) {
+    return apiError(req, res, {
+      status_code: 403,
+      api_error: {
+        type: "workspace_auth_error",
+        message:
+          "Only users that are `admins` for the current workspace can retrieve its monthly usage.",
+      },
+    });
+  }
+
+  switch (req.method) {
+    case "GET":
+      const analytics = await getAnalytics(req.query.wId as string);
+      res.status(200).json(analytics);
+      return;
+
+    default:
+      return apiError(req, res, {
+        status_code: 405,
+        api_error: {
+          type: "method_not_supported_error",
+          message: "The method passed is not supported, GET is expected.",
+        },
+      });
+  }
+}
+
+export default withLogging(handler);
+
+interface MemberCountQueryResult {
+  member_count: number;
+}
+
+interface ActiveUsersQueryResult {
+  last_7_days_active_users: number;
+  wow_growth_pct: number;
+  last_30_days_active_users: number;
+  mom_growth_pct: number;
+}
+
+interface AverageWeeklyDailyActiveUsersQueryResult {
+  last_7_days_average_daily_active_users: number;
+  wow_growth_pct: number;
+}
+
+async function getAnalytics(
+  wId: string
+): Promise<GetWorkspaceAnalyticsResponse> {
+  const [memberCountResults, activeUsersResult, averageWeeklyDauResult] =
+    await Promise.all([
+      front_sequelize.query<MemberCountQueryResult>(
+        `
+      SELECT COUNT(DISTINCT "memberships"."userId") AS "member_count"
+      FROM "memberships"
+      JOIN "workspaces" ON "memberships"."workspaceId" = "workspaces"."id"
+      WHERE "workspaces"."sId" = :wId
+      AND "memberships"."role" <> 'revoked';
+    `,
+        {
+          replacements: {
+            wId,
+          },
+          type: QueryTypes.SELECT,
+        }
+      ),
+      front_sequelize.query<ActiveUsersQueryResult>(
```

Comment:
big queries like this are fine while you test but imho should be refactored as something readable once you're clear
Was it your intention? if needed, I can prodvide suggestions

Reply:
As per IRL, aligned, will try to split them a bit more